### PR TITLE
Make OC.Features optional to non-default tenant

### LINKED ISSUE #12723: Make OrchardCore.Features not always enabled
### Is your feature request related to a problem? Please describe.
Currently, the module `OrchardCore.Features` is always enabled. We recently added a feature for the default tenant to be able to manage features on the behalf of any tenant. Can we remove the `AlwaysEnabled` flag from `OrchardCore.Features` as the saas owner may not want to allow the tenant to control their own features.

### PR DESCRIPTION
**Pull request by MikeAlhayek**
Fix #12723

### CODE REVIEW

**Comment by Skrypt** [comment]
More on this. I think we should stop using the ShellSettings.IsDefaultShell() and rely on a TenantId with a feature that says that this specific tenant is a tenant management one. Hence why I started a PR that introduces a TenantId.

**Comment by MikeAlhayek** [comment]
> More on this. I think we should stop using the ShellSettings.IsDefaultShell() and rely on a TenantId with a feature that says that this specific tenant is a tenant management one. Hence why I started a PR that introduces a TenantId.

@Skrypt this PR isn't using `ShellSettings.IsDefaultShell()`

**Comment by sebastienros** [comment]
Migration documentation: modules that use the recipe migrations need to depend on `OrchardCore.Recipes.Core`.

**Comment by jtkech** [comment]
I tagged it as `don't merge` but only temporarily just because I want to review it before ;)

For example in a manifest we can't have a dependency on a core project which is not a module.

I can't right now but I'll try to review it this night

**Comment by MikeAlhayek** [comment]
@jtkech we'll await your approval before we merge it. The only reason we can list `OrchardCore.Recipes.Core` in the manifest file is because a new feature was added with the id `OrchardCore.Recipes.Core`. As you can see this feature is `EnabledByDependencyOnly` which should be enable/disabled on demand as needed to provide the core services and prevent duplicate service registrations. 

```
[assembly: Feature(
    Id = "OrchardCore.Recipes.Core",
    Name = "Recipes",
    Description = "Provides recipe services.",
    Dependencies = new[]
    {
        "OrchardCore.Recipes.Core"
    },
    Category = "Infrastructure",
    EnabledByDependencyOnly = true
)]
```

**Comment by jtkech** [comment]
Ah okay, I will look at it this night.

Hmm but looks wrong that `OrchardCore.Recipes.Core` depends on itself.

**Comment by MikeAlhayek** [comment]
@jtkech it looks wrong because it is wrong. nice catch... I was just testing you ;)

I also added the documentation for the 1.6 release as @sebastienros requested.

**Comment by jtkech** [comment]
Okay cool !!!

**Comment by MikeAlhayek** [comment]
@jtkech were you able to valid this PR?

**Comment by jtkech** [comment]
@MikeAlhayek Oops sorry, no problem I will look at it this night.

File: src/OrchardCore.Modules/OrchardCore.Recipes/Startup.cs [hunk 1]
```diff
@@ -57,4 +55,18 @@ public override void Configure(IApplicationBuilder builder, IEndpointRouteBuilde
...
+    public class RecipesCoreStartup : StartupBase
+    {
+        public override void ConfigureServices(IServiceCollection services)
+        {
+            services.AddScoped<IRecipeHarvester, ApplicationRecipeHarvester>();
+            services.AddScoped<IRecipeHarvester, RecipeHarvester>();
+            services.AddTransient<IRecipeExecutor, RecipeExecutor>();
+            services.AddScoped<IRecipeMigrator, RecipeMigrator>();
+            services.AddScoped<IRecipeReader, RecipeReader>();
+            services.AddScoped<IRecipeEnvironmentProvider, RecipeEnvironmentFeatureProvider>();
```
[the remark this answers is not in the document]
Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Recipes/Startup.cs:69]: Because reusing it cases issues. If one needs the services, they should depend on the core project.

File: src/OrchardCore.Modules/OrchardCore.Features/Controllers/AdminController.cs [hunk 2]
```diff
@@ -99,6 +101,11 @@ public async Task<ActionResult> Features(BulkActionViewModel model, bool? force,
         [HttpPost]
         public async Task<IActionResult> Disable(string id, string tenant)
         {
+            if (String.IsNullOrEmpty(id))
+            {
+                return NotFound();
```
[the remark this answers is not in the document]
Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Features/Controllers/AdminController.cs:106]: I think `String.IsNullOrEmpty(id)` is safer here.... but I'll change it to `== null` instead

File: src/docs/releases/1.6.0.md [hunk 3]
```diff
@@ -0,0 +1,9 @@
+# Orchard Core 1.6.0
+
+Release date: Not yet released
+
+## Breaking Changes
+
+### Features and Recipes
+
+The `OrchardCore.Features` module is no longer always enabled. It used to depend on `OrchardCore.Recipes` and is no longer the case. If you find yourself using recipes services like `IRecipeMigrator` in your module, then be sure to add dependency to `OrchardCore.Recipes.Core` which ensures that the recipe services are available before trying to use them.
```
[the remark this answers is not in the document]
Comment (MikeAlhayek) [reply src/docs/releases/1.6.0.md]: @jtkech  I made some changes here. If you have something specific you want changes, please suggest it and I'll merge it.

**Comment by MikeAlhayek** [comment]
@jtkech I made some updates as per your request. If you are good with it, either merge or approve please.

**Comment by jtkech** [comment]
@MikeAlhayek Okay, just did an additional little review.

As said in the review, in fact the related modules using the recipe migrator were already missing a dependency on OC.Recipes, and seems to not be related to this PR, not saying that it is not good to be fixed here but just to well understand the goal of this PR.

So, even if core modules is a good idea (we may need to discuss a little more on it), for the goal of this PR it seems that we only need to remove `OC.Features` from the dependencies of `OC.Recipes`.

Then we would just need to add or not `OC.Features` in the recipes of other non Default tenants, If you want the same recipe for the Default tenant, `OC.Features` can be pre-configured for it only.

Note: Never tried so not sure but maybe we can also use Feature Profiles to filter `OC.Features`.

---

So in summary, unless some tweaks as missing null checks, maybe this PR only needs to remove one line, the dependency of `OC.Recipes` on `OC.Features` that doesn't seem to be required, and maybe also add the already missing dependency of some modules on `OC.Recipes`.

---

Hmm, the above assuming that we don't change the behavior of `OC.Features` being an `IsAlwaysEnabled` feature, meaning that it can't be disabled once enabled.

Yes because, even if here we keep the same behavior for the Default tenant, not sure it's good that this is no more the case for the other tenants, It may be unexpected that we now can disable `Features` through the Admin UI, particularly on existing non Default tenants.

**Comment by MikeAlhayek** [comment]
@jtkech I think I went down the path of making Recipes.Core because a service depends on recipes that isn't available unless Features in enabled. I don't recall the exact service name now. 

Either way, I don't think there is a good reason to undo some of these good changes that already implemented in this PR. With this PR, we can even make Recipes not always enabled if we want to hide the UI. In theory, everything should still work, as long as everything that needs core recipe services depends on the new core feature.

File: src/docs/releases/1.6.0.md [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+
+Release date: Not yet released
+
+## Breaking Changes
+
+### Features and Recipes
+
+The `OrchardCore.Features` module is no longer always enabled. It used to depend on `OrchardCore.Recipes` feature and is no longer the case. If you find yourself using recipes services like `IRecipeMigrator` in your module, then be sure to add dependency to `OrchardCore.Recipes.Core` feature in your module's Manifest' file to ensures that the recipe services are available before trying to use them.
+
+If your are using setup recipes and want to enable `Features`, be sure to explicitly enable `OrchardCore.Feature`
```
[the remark this answers is not in the document]
Comment (MikeAlhayek) [reply src/docs/releases/1.6.0.md]: I made changes to the docs.

**Comment by jtkech** [comment]
> I think I went down the path of making Recipes.Core because a service depends on recipes that isn't available unless Features in enabled. I don't recall the exact service name now.

Maybe another missing dependency, anyway I'm not against core features, just to well understand.

> Either way, I don't think there is a good reason to undo some of these good changes that already implemented in this PR.

I like this sentence ;)

> With this PR, we can even make Recipes not always enabled if we want to hide the UI.

Okay that's the other reason not directly related to this PR that I wanted to be confirmed, hmm we already discussed about it If I remember correctly.

---

Okay so all looks good, the only remaining thing is about the `OC.Features` that would not be anymore an `IsAlwaysEnabled` (can't be disabled once enabled) feature for non Default tenants.

So for this last thing allowing for a non default tenant to disable the `OC.Features` through the admin UI, I'm not against (I"m against nothing ;)) but I let others decide, e.g. @sebastienros.

**Comment by MikeAlhayek** [comment]
@jtkech I agree with you. I think if the feature is enabled, only Default should be able to disable it. This way if the Default tenant enabled it for other tenant, they won't be able to accidentally disable it and lose the ability to manage features. I'll make this change

File: src/docs/releases/1.6.0.md [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+
+Release date: Not yet released
+
+## Breaking Changes
+
+### Features and Recipes
+
+`OrchardCore.Features` no longer auto enabled by `OrchardCore.Recipes` as it doesn't depend on it anymore. If you find yourself using recipes services like `IRecipeMigrator` in your module, be sure to add dependency to `OrchardCore.Recipes.Core` feature in your module's Manifest file to ensures that the recipe services are available before using them.
+
+If your are using setup recipes and want to enable `Features`, be sure to explicitly enable `OrchardCore.Feature` feature.
```
[the remark this answers is not in the document]
Comment (MikeAlhayek) [reply src/docs/releases/1.6.0.md:11]: Done

**Comment by MikeAlhayek** [comment]
@jtkech I made that changes. Now, only host can enable/disable feature module for others. No tenant can disable their own `Features` module.

**Comment by sebastienros** [comment]
If any feature can't be enabled by a tenant, it shouldn't be able to disable it either.

**Comment by MikeAlhayek** [comment]
@sebastienros I made the requested change here. The only issue I am seeing from a UI perspective here, if a site admin disable their own Features module, they'll get a 404 on the redirect. So instead of the 404, I redirect the user to the dashboard if they disabled their own features UI.

@jtkech Are you good with this PR or do you have any change to add?

**Comment by jtkech** [comment]
@MikeAlhayek Okay no problem but there is still something that I don't understand.

So now if a non `Default` tenant disable its own `OC.Features` it is redirected to the dashboard, okay good, but it seems that it is still allowed to do it.

I would have made another distinction in the view model, not only to know if this is the `Default` tenant, but also if it is rendered by itself. If by itself it would not be allowed to disable its own  `OC.Features`, hmm looks like this same first condition could be also applied to the `Default` tenant.

Then if not managed by itself for now it means that it is by the `Default` tenant, but maybe safer to still check this other state before finally allowing to disable `OC.Features` of another tenant.

**Comment by MikeAlhayek** [comment]
@jtkech Thanks for your feedback.

> So now if a non Default tenant disable its own OC.Features it is redirected to the dashboard, okay good, but it seems that it is still allowed to do it.

Yes this is allowed as per @sebastienros request. He said if they disable their own `Features` feature, no problem for us.

> I would have made another distinction in the view model, not only to know if this is the Default tenant, but also if it is rendered by itself. If by itself it would not be allowed to disable its own OC.Features, hmm looks like this same first condition could be also applied to the Default tenant.

This is the exact behavior we have. Maybe we should rename the new `IsDefaultTenant`? I can't think of a good name.

> Then if not managed by itself for now it means that it is by the Default tenant, but maybe safer to still check this other state before finally allowing to disable OC.Features of another tenant.

Now sure that I understand. In the controller we check to see if the current tenant can disable the feature.

File: src/OrchardCore.Modules/OrchardCore.Features/Controllers/AdminController.cs [hunk 6]
```diff
@@ -63,6 +67,7 @@ public async Task<ActionResult> Features(string tenant)
                 // if the user provide an invalid tenant value, we'll set it to null so it's not available on the next request
                 tenant = settings?.Name;
                 viewModel.Name = settings?.Name;
+                viewModel.IsDefaultTenant = _shellSettings.IsDefaultShell() && (settings == null || settings.IsDefaultShell());
```
Comment (jtkech) [inline src/OrchardCore.Modules/OrchardCore.Features/Controllers/AdminController.cs:70]: Would be better to init a `var shellSettings = _shellSettings;` before and use it in the delegate of the child scope.

But even if `_shellSettings` is a tenant singleton here we may be in a different tenant container, so better to not capture `_shellSettings` belonging to another container. So maybe just init before `var name = _shellSettings.Name` or even more `var isDefaultShell = _shellSettings.IsDefaultShell()`, and then use it in the delegate.

Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Features/Controllers/AdminController.cs]: @jtkech done

**Comment by jtkech** [comment]
> Yes this is allowed as per @sebastienros request. He said if they disable their own Features feature, no problem for us.

Okay I'm not against but this is to prevent this "issue" that we made this feature `IsAlwaysEnabled` which is no more the case, maybe worth to be confirmed, e.g. in case If it was just to make things simpler. 

Easy to keep the `IsAlwaysEnabled` behavior (can't be disabled once enabled) except when this is the Default tenant that manages another tenant.

- If Manager == Managed => can't disable the Features feature.

- If Manager != Managed AND Manager is the Default tenant => can disable the Features feature.

**Comment by MikeAlhayek** [comment]
Yes I do not disagree with you. This is what I had done. So basically Features will be always enabled for tenants once enabled. Only Default tenant will be able to disable it using the Features UI. If you watch the recording for last meeting on Tuesday, you’ll see his remarks. 

The only issue with allowing the tenant to disable their own Features tenant, is that they can’t enable it again from the UI. They will be able to enable it via recipes is the deployment feature is enabled. Either way, let us know what do you recommend to do here and we can discuss with your feedback with @sebastienros

**Comment by jtkech** [comment]
Okay then, as said I'm open to any suggestion, just to be sure there was no confusion.

otherwise I would recommend to keep the previous behavior and only allow the Default tenant to make new things needed in this PR.

Hmm maybe @sebastienros doesn't want a feature which is no more `IsAlwaysEnabled` but that a tenant is still not able to disable for itself. Hmm but here still the case for the Default tenant on itself.

Hmm, maybe better to keep `OC.Features` as `IsAlwaysEnabled` (less change on what was before) and only allow the `Default` tenant to violate the disabling restriction when acting on another tenant.

**Comment by MikeAlhayek** [comment]
@sebastienros thoughts here? It sounds like your thoughts are the opposite of what is @jtkech is suggesting here. We just need a way to be able to disable Features for tenants.

File: src/OrchardCore.Modules/OrchardCore.Features/Controllers/AdminController.cs [hunk 7]
```diff
@@ -58,11 +62,14 @@ public async Task<ActionResult> Features(string tenant)
...
             var viewModel = new FeaturesViewModel();
 
+            var isDefaultTenant = _shellSettings.IsDefaultShell();
+
             await ExecuteAsync(tenant, async (featureService, settings) =>
             {
                 // if the user provide an invalid tenant value, we'll set it to null so it's not available on the next request
                 tenant = settings?.Name;
                 viewModel.Name = settings?.Name;
+                viewModel.IsDefaultTenant = isDefaultTenant && (settings == null || settings.IsDefaultShell());
```
Comment (jtkech) [inline src/OrchardCore.Modules/OrchardCore.Features/Controllers/AdminController.cs:72]: What you could do is to always pass a non null settings in your delegate even if you didn't create a tenant scope (acting on the current tenant), this would simplify the check => `isDefaultTenant && settings.IsDefaultShell()`

Replacing

    // at this point we apply the action on the current tenant
     await action(new FeatureService(_shellFeaturesManager, _extensionManager), null);

with

    await action(new FeatureService(_shellFeaturesManager, _extensionManager), _shellSettings);

Also under `shellScope.UsingAsync()`.

    // at this point we apply the action on the given tenant
    await action(new FeatureService(shellFeatureManager, extensionManager), settings);

We could do

    await action(new FeatureService(shellFeatureManager, extensionManager), scope.ShellContext.Settings);

Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Features/Controllers/AdminController.cs:72]: Sure. I'll make these changes tomorrow. But I we need to agree on the other item

Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Features/Controllers/AdminController.cs]: I made the requested change.

**Comment by jtkech** [comment]
@sebastienros @MikeAlhayek 

Just saw the last meeting, yes we can set a feature as `IsAlwaysEnabled` in a manifest file, it means that you can enable it but not disable it once enabled, yes I don't like the name but this is an existing setting.

It is only used by the UI and not managed by the infrastructure, so yes I think we can disable it through a recipe, but it was introduced to prevent people from breaking their tenant/site from the UI. For info we also have `AlwaysEnabled` features which are managed by the infrastructure but this is not the same.

And the `Features` feature has been marked as `IsAlwaysEnabled` to fix an issue opened at some point, but here this is no more the case to allow the Default tenant to disable it for other tenants.

So the goal (not saying it is good) was to keep the same `IsAlwaysEnabled` behavior for all tenants (including the Default) when they are managing their own features, even if here`Features` is no longer marked as `IsAlwaysEnabled`, then only allow to disable this feature when a tenant is managing the features of another tenant (here for now can only be done by the Default tenant).

I'm not against anything but maybe keep the exact same behavior as before and then just allow the Default tenant to disable `Features` for other tenants.

Hmm, IMHO I think it is better to keep `Features` marked as `IsAlwaysEnabled` as it may have been done for a good reason, then just allow the disable button in the UI when we know that we are in the context (state retrieved from the model) of the Default tenant managing features of another tenant, which for now here can be resumed by: "a tenant is managing the features of another tenant".

**Comment by jtkech** [comment]
@MikeAlhayek 

Yes during the meeting @sebastienros was saying that it is not good to manage differently a given feature as `OC.Features` and that we would need a setting for that, okay I understand but we already have such a setting `IsAlwaysEnabled` and that is used for the `Feature` feature .

So we just need a confirmation, maybe he was not remembering about the `IsAlwaysEnabled` setting.

Anyway, as done currently where `Features` is no more `IsAlwaysEnabled`, we still need a special behavior to still not allow the `Default` tenant to disable it for himself (only others would be able to do that).

So my recommendation is to keep `OC.Features` as `IsAlwaysEnabled` and just allow the disable buton as a special behavior but only in the new special context of a [Default] tenant managing another tenant.

**Comment by jtkech** [comment]
@MikeAlhayek 

I think we can safely re-use `IsAlawaysEnabled` for `OC.Features` because it would be as before.

If @sebastienros really wanted to change this existing behavior, will be easy to tweak again.

Then only allow the new behavior (the disable button) for the Default tenant when it manages another tenant, easy to add this state to the model, as I remember by checking if the shellSettings are different.

Then I will let you re-tried it a little, then I will approve and merge.

**Comment by MikeAlhayek** [comment]
@jtkech so to ensure I understand the request, you want to remove the IsAlwaysEnabled on the feature. However, a tenant should not be able to disable it on their own from their UI. A default tenant will still be able to enable/disable the feature for any other tenant. Correct? I'll try work on this tomorrow.

**Comment by jtkech** [comment]
@MikeAlhayek 

> you want to remove the IsAlwaysEnabled on the feature.

No, you already removed ` IsAlwaysEnabled = true` from `OC.Features`, I suggest to re-use it instead, even if at the end we will have the same desired result.

> However, a tenant should not be able to disable it on their own from their UI. A default tenant will still be able to enable/disable the feature for any other tenant. Correct?

Yes the goal will be to preserve this desired behavior, we will just have to change the view model state we pass to the view that will check it to allow the Disable button, this in a simpler way than before I think.

**Comment by MikeAlhayek** [comment]
If we set the IsAlwaysEnabled back to true, I don't think the default tenant will be able to disable the feature. I'll try it

**Comment by jtkech** [comment]
Do you mean it is not only checked by the UI? Let me check it.

It seems that `IsAlwaysEnabled` is only used by `FeatureService.GetModuleFeaturesAsync()` that is called by the Features Controller, so normally checking our state in the view would be okay.

**Comment by MikeAlhayek** [comment]
@jtkech you are right. `IsAlwaysEnabled` is just used by the UI. I made the requested changed. This is what we have now
1. `OrchardCore.Features` has `IsAlwaysEnabled` flag set.
2. If a Default tenant is executing on the behalf of other tenants, they are able to disable and re-enable `Features` feature for that tenant.
3. No tenant can disable `Features` module for their own instance.

I tested this out and all seems to be doing exactly what we want. Feel free to Merge or at least approve it and I'll merge it.
